doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+go to the slow path to invoke the call. Afterwards, the class `C` will be allowed to be loaded and to enter the
+system. When we do nopable guard patching, the guard changes the state. We convert an unconditional fall through
+into an unconditional goto.  Even if it is the object `B` that flows through, it will still jump to the slow
+path to invoke the virtual call. Nopable guards require JIT to monitor the changes of program behaviour such as
+class loading/unloading and do the patching to ensure inlined code is not invoked when we are not allowed to.
+
+## Explicit Guard
+
+Explicit guards are the guards that we generate the code for what we have in the trees.
+
+There are two different cases. One case is that there is [a bit in every J9Method](https://github.com/eclipse-openj9/openj9/blob/8393a7277cbd664acc8c9477d66b2dc9a9d6f412/runtime/compiler/env/j9method.cpp#L5074-L5077)
+indicating whether the method is overridden or not.  Any time a subclass is loaded, the VM checks if the method
+in that class overrides the method in the super class and it sets that bit for that method. We can always
+inline the method with a guard that checks the method. We explicitly test that bit in our IL trees. If it is not
+overridden, the inline code is executed, otherwise it invokes the virtual call.
+
+Another case of explicit test is based on profiling which is more common by default. Profiler tracks what type
+of objects we encounter when we have the virtual call or the interface call. That is the receiver type.
+The receiver is the object that is receiving the message to invoke the call. Either the interpreter or profiler
+or JProfiler keeps track of the type of an object.
+
+## Conditions Where A Nopable Guard Cannot Be Used
+
+If there is more than one target based on the class hierarchy and the state of the program, an explicit guard
+must be used. Java is a dynamic language and classes can be loaded in the future. Nopable guard can only be used
+if there is one and only target regarding to the current state of the program or the current class hierarchy
+when the compilation is done.
+
+For direct calls there is exactly one possibility. When a method is invoked on an abstract class, the object that
+will receive the method cannot be the type of the abstract class, because abstract class cannot be instantiated.
+It will be a subclass of the abstract class. If there is only one subclass `A` of the abstract class `B`, there
+is no point of checking if the object you use at the location is `B` or not. `A` being `B` is the only
+possibility that exists right now.
+
+When HCR is disabled, we know which method will be called and the bytecode within that method will not change.
+With HCR, we know which method will be called but the bytecode within that method can change. Therefore, we
+still need to have an HCR guard that is nopable.
+
+For virtual calls, we do not know what to call, it could be `A.foo()` or `B.foo()` or `C.foo()`. We only know it
```

Comment:
Would this discussion be better around the place where abstract and interface calls are discussed ? As opposed to after we discuss HCR for a bit...